ui/plugin-system/src/components/PluginRegistry/plugin-indexes.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useEvent } from '@perses-dev/core';
+import { useCallback, useRef } from 'react';
+import { PluginMetadata, PluginModuleResource, PluginType } from '../../model';
+
+export type GetInstalledPlugins = () => Promise<PluginModuleResource[]>;
+
+export interface PluginIndexes {
+  // Plugin resources by plugin type and kind (i.e. look up what module a plugin type and kind is in)
+  pluginResourcesByTypeAndKind: Map<string, PluginModuleResource>;
+  // Plugin metadata by plugin type
+  pluginMetadataByType: Map<string, PluginMetadata[]>;
+}
+
+/**
+ * Returns an async callback for getting indexes of the installed plugin data.
+ */
+export function usePluginIndexes(getInstalledPlugins: GetInstalledPlugins) {
+  // Creates indexes from the installed plugins data (does useEvent because this accesses the getInstalledPlugins prop
+  // and we want a stable reference for the callback below)
+  const createPluginIndexes = useEvent(async (): Promise<PluginIndexes> => {
+    const installedPlugins = await getInstalledPlugins();
```

Comment:
do we need to catch any errors here?

Reply:
Good question: I decided to let errors just bubble up to the call site for now (i.e. wherever `usePlugin` or `useListMetadata` are called). But totally possible we want to report this once via some "global" mechanism and then let it bubble up. We may need to think about having some plugin runtime error reporting context that folks could implement when embedding (i.e. "use a snackbar" or soemthing)...